### ISSUE TITLE
UPGRADE FAILED: failed to replace object: PersistentVolumeClaim "pvc-0" is invalid: spec: Forbidden: is immutable after creation except resources.requests for bound claims

### ISSUE THREAD

**Issue by magick93** [helm/helm#9861]
# Background

I have a stateful set that has a single persistent volume.

I need to modify the command from:

```
containers:
        - name: chain-node
          imagePullPolicy: IfNotPresent
          image: {{ .Values.image }}
          command: [ "/bin/cat" ]
          tty: true
          stdin: true
```
to:
```
containers:
        - name: chain-node
          imagePullPolicy: IfNotPresent
          image: {{ .Values.image }}
          command: ["/start"]
          args: [{{ printf "--%s" .Values.network.name }}]
```

Thats it. No changes to the volumes. 

I then run `helm upgrade stellar . --namespace stellar --force` and get the following error:

```
Error: UPGRADE FAILED: failed to replace object: PersistentVolumeClaim "stellar-pvc-stellar-0" is invalid: spec: Forbidden: is immutable after creation except resources.requests for bound claims
```
The changes I've made are, as mentioned, to the command, not to the PVC, so helm shouldnt be trying to replace this.

This seems like, if I use a column, I cant change the command. That would seem extremely odd!

Output of `helm version`:

version.BuildInfo{Version:"v3.6.1", GitCommit:"61d8e8c4a6f95540c15c6a65f36a6dd0a45e7a2f", GitTreeState:"clean", GoVersion:"go1.16.5"}

Output of `kubectl version`:

Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"6b1d87acf3c8253c123756b9e61dac642678305f", GitTreeState:"clean", BuildDate:"2021-03-18T01:10:43Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.15-eks-cf75c9", GitCommit:"cf75c9601ec63a63ba54c1c47caa9945af1f3a81", GitTreeState:"clean", BuildDate:"2021-02-13T00:14:16Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 

Amazon EKS.

# Update

I reverted the `command` changes, and this time just made a super simple change to the `metadata.name`:

```
metadata:
  name: {{ .Release.Name }}-helloworld
```

I get the same error.

My expectation is that data in the PV is persistent. But I should be able to change commands, names etc. Persistent *volume* not persistent everything. 

# Workaround

I was able to move passed this issue by :
1. `helm template mychart . > output.yml` 
2. extract the statefulset yml to a separate file - `statefulset.yml`
3. `kubectl apply -f statefulset.yml`

**Comment by bacongobbler** [comment]
Interesting. This could be part of Kubernetes’ update logic under the covers causing this. Let us know if you can determine what causes this. I would suggest looking into pkg/kube for more information - that package interacts directly with Kubernetes client APIs.

**Comment by CharlieReitzel** [comment]
> Interesting. This could be part of Kubernetes’ update logic under the covers causing this. Let us know if you can determine what causes this. I would suggest looking into pkg/kube for more information - that package interacts directly with Kubernetes client APIs.

I am working on `helm upgrade` for an application (release = `query-server`, namespace = `qs`) and am currently blocked by this issue.   I found that if I generated the manifest for the installed release using the command:
```
helm get manifest -n qs query-server > qs-installed.manifest.yaml
```

Then capture the manifest for the upgrade using the command:
```
helm upgrade -n qs --dry-run query-server query-server > qs-upgrade.manifest.yaml
```

When I compare the two manifests, there are zero differences in the `PersistentVolumeClaim` objects, as expected.   But I also noticed, finally, a bug in my template.  I refer to a non-existent value name (typo in the name).  This resulted in no values for a couple PVC attributes (`volumeName`, `volumeMode`).    Once I fixed the typos, I no longer experienced this issue.

Long story short: It's probably worth a helm warning if a template renders no value at all for any given element.   As it is, I found the error message is difficult to understand.  :--)

I'm not sure about the scope, tho.  I doubt, somehow, the issue is limited to `kind: PersistentVolumeClaim`.  There may be other K8s objects which have similar update semantics.   Maybe best applied on an object-by-object basis.  I.e. create a method that works for any `kind` of object, but apply it only to to those known to exhibit this weirdness during `helm upgrade`.
